### PR TITLE
apps.plugin improvements

### PR DESCRIPTION
**Pull request by ktsaou**
- [x] apps.plugin now supports simple patterns when an asterisk is in the middle of a match
- [x] expanded kernel threads matching to group them into meaningful entities
- [x] when there is an asterisk in the middle of the command name, apps.plugin now uses a SIMPLE PATTERN to match the process
- [x] `apps_groups.conf` now categorizes useful kernel threads.
- [x] apps.plugin now extracts the full `comm` name from the `cmdline`.
- [x] apps.plugin now extract the service names.
- [x] fixed a bug is spawn_server_windows, where the opening quote was missing when the full path of the command line has a space in it.
- [x] fixed infinite loop in health, when `alarm-notify.sh` cannot be executed (`spawn_server_exec()` returns `NULL`).

### CODE REVIEW

File: src/collectors/apps.plugin/README.md [hunk 1]
```diff
@@ -132,6 +132,26 @@ its CPU resources will be cut in half, and data collection will be once every 2
...
+
+Process managers are configured in `apps_groups.conf` with the prefix `managers:`, like this:
+
+```
+managers: process1 process2 process3
+```
+
+Multiple lines may exist, all starting with `managers:`.
+
+The process names given here should as the operating system gives them. In Linux these
```
Comment (Ancairon) [inline src/collectors/apps.plugin/README.md:149]:
```suggestion
The process names given here should be exactly as the operating system sets them. In Linux these
```

File: src/collectors/apps.plugin/README.md [hunk 2]
```diff
@@ -140,48 +160,39 @@ group: process1 process2 ...
 
 Each group can be given multiple times, to add more processes to it.
 
-For the **Applications** section, only groups configured in this file are reported.
-All other processes will be reported as `other`.
-
-For each process given, its whole process tree will be grouped, not just the process matched.
-The plugin will include both parents and children. If including the parents into the group is
-undesirable, the line `other: *` should be appended to the `apps_groups.conf`.
+For each process given, all its sub-processes will be grouped, not just the process matched.
```
Comment (Ancairon) [inline src/collectors/apps.plugin/README.md:163]:
```suggestion
For each process given, all of its sub-processes will be grouped, not just the matched process.
```

File: src/collectors/apps.plugin/README.md [hunk 3]
```diff
@@ -140,48 +160,39 @@ group: process1 process2 ...
...
 
 The process names are the ones returned by:
 
--   `ps -e` or `cat /proc/PID/stat`
--   in case of substring mode (see below): `/proc/PID/cmdline`
+-   **comm**: `ps -e` or `cat /proc/{PID}/stat`
+-   **cmdline**: in case of substring mode (see below): `/proc/{PID}/cmdline`
+
+On Linux **comm** is limited to just a few characters. `apps.plugin` attempts to find the entire
+**comm** name by looking for it at the **cmdline**. When this succeeds, the entire process name
```
Comment (Ancairon) [inline src/collectors/apps.plugin/README.md:171]:
```suggestion
**comm** name by looking for it at the **cmdline**. When this is successful, the entire process name
```

File: src/collectors/apps.plugin/README.md [hunk 4]
```diff
@@ -140,48 +160,39 @@ group: process1 process2 ...
...
--   `*name` _suffix_ mode: will search for processes ending with `name` (at `/proc/PID/stat`)
--   `name*` _prefix_ mode: will search for processes beginning with `name` (at `/proc/PID/stat`)
--   `*name*` _substring_ mode: will search for `name` in the whole command line (at `/proc/PID/cmdline`)
+-   `*name` _suffix_ mode: will match a **comm** ending with `name`.
+-   `name*` _prefix_ mode: will match a **comm** beginning with `name`.
+-   `*name*` _substring_ mode: will search for `name` in **cmdline**.
 
-If you enter even just one _name_ (substring), `apps.plugin` will process
-`/proc/PID/cmdline` for all processes (of course only once per process: when they are first seen).
+Asterisk may appear in the middle of `name` (like `na*me`), without affecting what is being
```
Comment (Ancairon) [inline src/collectors/apps.plugin/README.md:183]:
```suggestion
Asterisks may appear in the middle of `name` (like `na*me`), without affecting what is being
```

File: src/collectors/apps.plugin/README.md [hunk 5]
```diff
@@ -140,48 +160,39 @@ group: process1 process2 ...
...
 To add processes with double quotes, enclose them in single quotes: `'process with this " double quote'`
 
-If a group or process name starts with a `-`, the dimension will be hidden from the chart (cpu chart only).
-
-If a process starts with a `+`, debugging will be enabled for it (debugging produces a lot of output - do not enable it in production systems).
-
-You can add any number of groups. Only the ones found running will affect the charts generated.
-However, producing charts with hundreds of dimensions may slow down your web browser.
-
 The order of the entries in this list is important: the first that matches a process is used, so put important
```
Comment (Ancairon) [inline src/collectors/apps.plugin/README.md:190]:
```suggestion
The order of the entries in this list is important: the first one that matches a process is used, so follow a top-down hierarchy.
```

File: src/collectors/apps.plugin/README.md [hunk 6]
```diff
@@ -140,48 +160,39 @@ group: process1 process2 ...
...
-ones at the top. Processes not matched by any row, will inherit it from their parents or children.
-
-The order also controls the order of the dimensions on the generated charts (although applications started
-after apps.plugin is started, will be appended to the existing list of dimensions the `netdata` daemon maintains).
+ones at the top. Processes not matched by any row, will inherit it from their parents.
```
Comment (Ancairon) [inline src/collectors/apps.plugin/README.md:191]:
```suggestion
Processes not matched by any row, will inherit it from their parents.
```

@ktsaou here I remove the remnant of the previous sentence, but I got a question:

> Processes not matched by any row, will inherit **it** from their parents.

What does "it" represent here? Could it be referring to the snippet you removed:

```md
If a group or process name starts with a `-`, the dimension will be hidden from the chart (cpu chart only).

If a process starts with a `+`, debugging will be enabled for it (debugging produces a lot of output - do not enable it in production systems).

You can add any number of groups. Only the ones found running will affect the charts generated.
However, producing charts with hundreds of dimensions may slow down your web browser.
```

If it does, then probably the whole line can be deleted.

File: src/collectors/apps.plugin/README.md [hunk 7]
```diff
@@ -140,48 +160,39 @@ group: process1 process2 ...
...
 
-There are a few command line options you can pass to `apps.plugin`. The list of available options can be acquired with the `--help` flag. The options can be set in the `netdata.conf` file. For example, to disable user and user group charts you should set
+There are a few command line options you can pass to `apps.plugin`. The list of available
+options can be acquired with the `--help` flag. The options can be set in the `netdata.conf` file.
```
Comment (Ancairon) [inline src/collectors/apps.plugin/README.md:194]:
```suggestion
options can be acquired with the `--help` flag. The options can be set in the `netdata.conf` using the [`edit-config` script](/docs/netdata-agent/configuration/README.md).
```

File: src/collectors/apps.plugin/README.md [hunk 8]
```diff
@@ -140,48 +160,39 @@ group: process1 process2 ...
...
+For example, to disable user and user group charts you should set:
```
Comment (Ancairon) [inline src/collectors/apps.plugin/README.md:195]:
```suggestion
For example, to disable user and user group charts you would set:
```

**Review by Ancairon — COMMENTED** [review]
Read through the README, just some minor fixes here and there, only one real question regarding something that is vague in a sentence.

File: src/collectors/apps.plugin/README.md (hunk 6 above)
Comment (ktsaou) [reply src/collectors/apps.plugin/README.md:191]: Their group.
So we have PIDs (processes) which we group together in a meaningful way. So, PIDs not matches by any row, will inherit their group from their parents.

**Pull request merged**